### DISCUSSION TITLE
1주차 사전 과제 문주영

### DISCUSSION THREAD

**Discussion by Moon-ju-young** [MrMirror21/2024-KHU-CSE-TrackStudy-Web#21] [General]
### 1. 명명 규칙(Naming Convention)이라는 것이 있습니다. 어떤 규칙들이 있는지, 각 규칙들은 어떻게 코드를 작성하는지, 자신이 그동안 사용해왔던 명명 규칙은 무엇인지 적어주세요.

**대표적 방식**
- 카멜 케이스 (camelCase) : 첫 단어를 제외하고 단어의 첫 글자만 대문자로 작성
- 파스칼 케이스 (PascalCase) : 단어의 첫 글자만 대문자로 작성
- 스네이크 케이스 (snake_case) : 모든 문자를 소문자로 작성하고 단어 사이를 언더바(_)로 구분

**기타**
- 케밥 케이스 (kebab-case) : 모든 문자를 소문자로 작성하고 단어 사이를 대시(-)로 구분
- 플랫 케이스 (flatcase) : 모든 문자를 소문자로 작성
- 스크리밍 케이스 (SCREAMING CASE) : 모든 문자를 대문자로 작성

이 외에도 스크리밍 케이스와 스네이크 케이스가 결합된 스크리밍 스네이크 케이스 (SCREAMING_SNAKE_CASE)나 카멜 케이스와 스네이크 케이스가 합쳐서 만들어진 카멜 스네이크 케이스 (camel_Snake_Case) 등 케이스들이 합쳐서 만들어진 케이스들도 있다.

나는 주로 스네이크 케이스를 사용한다.

---

### 2. 함수가 순수하다는 것은 무엇을 의미할까요?

함수가 순수하다는 것은 함수에 부수 효과*가 없다는 것을 의미하며 같은 말로 동일한 인자가 주어졌을 때 항상 똑같은 값이 반환된다는 뜻이다.
\* 부수 효과 (side effect) : 함수의 인자 외에도 외부의 상태를 참조 및 변경하는 것과 같이 함수의 결과에 영향을 미치는 요인. 

---

### 3. 개발 시에 순수 함수(pure function)로만 모든 코드를 작성하는 것에 대한 생각을 적어주세요.

상태 변화를 추적하기 쉽고, 평가(실행) 시점이 무관하다는 점에서 함수를 활용하기 유용하고 오류를 잡아내기 쉬울거 같다. 모듈화에도 편할 것 같지만 부수효과가 필요할 때는 main code에 다 작성해야 한다는 점이 불편하지 않을까 싶다.

---

### 4. 자바스크립트에서 함수를 작성하는 문법은 다음과 같습니다. 이 외의 함수를 작성하는 문법 두 가지가 무엇이며, 어떤 때 사용하나요?

```jsx
function myFunc(param) {
	statements
}
```

위에서 작성한 방식은 함수 선언식이다. 함수 선언식 말고 함수 변수에 함수 참조값을 저장하는 방식인 함수 표현식으로도 작성할 수 있다. 함수 표현식은 함수 선언식과는 달리 호이스팅(hoisting, js 내부 알고리즘으로 함수나 변수를 먼저 선언하는 것을 말함)이 되지 않아 함수 표현식의 작성 위치에 따라 함수를 사용하지 못할 수 있다.

함수 표현식은 명명 함수 표현과 익명 함수 표현으로 나뉜다. 익명 함수 표현은 화살표 함수로도 표현할 수 있다.

* 명명 함수 표현
```jsx
const myFunc = function myFunction(param) {
	statements
};
```
* 익명 함수 표현
```jsx
const myFunc = function (param) {
	statements
};
```

* 화살표 함수

```jsx
const myFunc = (param) => {
	statements
};
```

---

#### 아래 파이썬 코드 속 같은 조건과 같은 결과를 자바스크립트로 출력하면 됩니다.

```python
>>> x = 1
>>> y = 2
>>> f"{x} + {y}는 {x + y}입니다."
'1 + 2는 3입니다.'
```

js에서는 가장 많이 쓰이는 방법은 템플릿 리터럴(template literal)을 통해 문자열 포맷팅을 구현한다. 템플릿 리터럴은 따옴표 말고 역음 부호(backtick, `)을 사용하여 문자열을 감싸고, 변수는 ${ } 으로 감싸 사용하면 된다.

```jsx
var x = 1;
var y = 2;
console.log(`${x} + ${y}는 ${x+y}입니다.`);
```

---

### 6. 다음과 같은 객체(object)에서 p 값을 출력하려면 o.p 와 같은 방식으로 출력해야 하는데, p 로 출력할 수 있는 방법은 없을까요?
```jsx
var o = { p: 42, q: true };

console.log(o.p); // 42
console.log(p); // p is undefined
```

객체의 속성을 변수에 할당한 후 그 변수를 사용할 수 있다. o.p 값을 변수 p에 할당한 다음 이 변수를 출력할 수 있다.

```javascript
var o = { p: 42, q: true };

var p = o.p;
console.log(o.p); // 42
console.log(p); // 42
```
이 방법은 객체의 속성을 새로운 변수에 할당하여 사용하므로 원하는 결과를 얻을 수 있다.

**Comment by MrMirror21** [comment]
답변해주신 사전 과제 잘 봤습니다.
전체적으로 좋은 퀄리티로 잘 답변해주셨네요. 스크리밍 케이스나 여러 케이스들이 합쳐져서 새로운 naming convention들이 발생한 건 저도 처음 알았습니다. 좋은 답변 감사합니다.
